discovery/snowflake/spcs.go

```diff
@@ -0,0 +1,440 @@
+	dsn, err := sf.DSN(&cfg)
+	if err != nil {
+		return nil, nil, fmt.Errorf("SPCS discovery plugin: failed to create DSN from Config. err: %w", err)
+	}
+
+	// Establish connection with snowflake.
+	db, err := sql.Open("snowflake", dsn)
+	if err != nil {
+		return nil, nil, fmt.Errorf("SPCS discovery plugin: failed to connect. err: %w", err)
+	}
+	defer db.Close()
+
+	// Execute query.
+	rows, err := db.Query(query)
+	if err != nil {
+		return nil, nil, fmt.Errorf("SPCS discovery plugin: failed to run query. %s, err: %w", query, err)
+	}
+	defer rows.Close()
+
+	// Get column names and types.
+	columns, err := rows.Columns()
+	if err != nil {
+		return nil, nil, fmt.Errorf("SPCS discovery plugin: error getting column names. err: %w", err)
+	}
+
+	// Create a slice to hold the values.
+	values := make([]interface{}, len(columns))
+	for i := range values {
+		var v interface{}
+		values[i] = &v
+	}
+
+	// Create a slice to store the rows.
+	var result [][]string
+
+	// Iterate through the rows.
+	for rows.Next() {
+		// Scan the current row into the values slice.
+		if err := rows.Scan(values...); err != nil {
+			return nil, nil, fmt.Errorf("SPCS discovery plugin: error scanning row. err: %w", err)
+		}
+
+		// Convert each value to a string and append it to the result slice.
+		rowValues := make([]string, len(columns))
+		for i := range columns {
+			if values[i] != nil {
+				// Convert the underlying value to a string.
+				strVal := fmt.Sprintf("%v", *values[i].(*interface{}))
+				rowValues[i] = strVal
+			}
+		}
+		result = append(result, rowValues)
+	}
+
+	// Check for errors during iteration.
+	if err := rows.Err(); err != nil {
+		return nil, nil, fmt.Errorf("SPCS discovery plugin: Error iterating over rows. err: %w", err)
+	}
+
+	return result, columns, nil
+}
+
+func getCurrentRole(rows [][]string) (string, error) {
+	// Expected to return a single row with one column.
+	for _, row := range rows {
+		role := row[0]
+		return role, nil
+	}
+	return "", fmt.Errorf("SPCS discovery plugin: error retrieving role.")
+}
+
+func getComputePoolsToScrape(rows [][]string, columns []string, role string) ([]string, error) {
+	// Create a slice to hold the computepools.
+	computePools := []string{}
+
+	// Iterate through the rows.
+	for _, row := range rows {
+		cpState, err := getColumnValue(row, columns, StateColumn)
+		if err != nil {
+			return nil, fmt.Errorf("SPCS discovery plugin: error retrieving compute pool state. %w", err)
+		}
+
+		ownerName, err := getColumnValue(row, columns, OwnerColumn)
+		if err != nil {
+			return nil, fmt.Errorf("SPCS discovery plugin: error retrieving compute pool owner name. %w", err)
+		}
+
+		cpName, err := getColumnValue(row, columns, NameColumn)
+		if err != nil {
+			return nil, fmt.Errorf("SPCS discovery plugin: error retrieving compute pool name. %w", err)
+		}
+
+		if strings.EqualFold(strings.ToLower(cpState), StateActiveValue) && strings.EqualFold(strings.ToLower(ownerName), strings.ToLower(role)) {
```

Comment:
What is the point of calling `ToLower` on both arguments to `EqualFold`?

Reply:
fixed.